### ISSUE TITLE
Juggernaut dies after a few days with "EMFILE, Too many open files"

### ISSUE THREAD

**Issue by gorman** [maccman/juggernaut#25]
After a couple of days of Juggernaut running (with moderate usage), it will die out with the following error:

---

net:891
        self.fd = socket(self.type);
                  ^
Error: EMFILE, Too many open files
    at net:891:19
    at Object.lookup (dns:138:5)
    at Stream.connect (net:886:9)
    at Timer.callback (/usr/local/lib/node/.npm/redis-client/0.3.5/package/lib/redis-client.js:852:29)
##     at node.js:757:9

As far as I can tell, sessions are properly being disconnected. I'm also only establishing one connection to Juggernaut per page load, so I'm not totally sure what I could optimize to keep this issue from happening. Any help is greatly appreciated!

**Comment by gorman** [comment]
I should mention that I'm running Ubuntu 10.04 and ulimit appears to be set to "unlimited" (some of my Googling returned suggestions of increasing the ulimit, but I don't think I need to?)

**Comment by maccman** [comment]
Interesting, can you try and pull from latest master and update the submodules to see if it's still happening?

**Comment by gorman** [comment]
Oops, I thought I had the latest deployed on the server -- evidently not :)

However, once I upgraded to the latest, I got this as soon as I tried to publish to Juggernaut:

buffer:67
      return this.utf8Write(string, offset);
                  ^
TypeError: Argument must be a string
    at Buffer.write (buffer:67:19)
    at [object Object].<anonymous> (vendor/node-static/lib/node-static.js:239:24)
    at [object Object].emit (events:27:15)
    at [object Object]._emitData (fs:735:10)
    at afterRead (fs:717:10)
    at node.js:757:9

I was able to get this working by changing:

buffer.write(chunk, offset, 0);

to:

chunk.copy (buffer, offset, 0);

I did this in vendor/node-static/lib/node-static.js. I copied that line from an old version of node-static.js and now everything is working fine. I'm not sure why I had to make that hack and I know it's not ideal, so I'm open to suggestions on that. I haven't had it running for long enough to see if the original problem is fixed though, but hopefully it is :) I'll report back.

Thanks for your help, I really appreciate all of the effort you put into Juggernaut and the community!
